rtla/cmd.sh

```diff
@@ -1,89 +1,264 @@
+# Validate each element in the array
+for e in "${events_arr[@]}"; do
+    # Validate each event against the desired regex pattern
+    if ! echo "$e" | grep -Pq "^[a-zA-Z0-9_]+(:[a-zA-Z0-9_]+)?$"; then
+        echo "Invalid event format: $e. It must be one word or two sets of words with a colon in between. Each word can include underscore '_' but no spaces."
+        exit 1
+    fi
 done
 
-uname=`uname -nr`
-echo "$uname"
-rpm -q rtla
 
-cpulist=`get_allowed_cpuset`
-echo "allowed cpu list: ${cpulist}"
+uname=`uname -nr` # get the kernel version
+echo "############# dumping env ###########" | storage
+env | storage
+echo "#####################################" | storage
 
-# change list seperators from comma to new line and sort it
-cpulist=`convert_number_range ${cpulist} | tr , '\n' | sort -n | uniq`
+echo " " | storage
+echo "########## container info ###########" | storage
+echo "/proc/cmdline:" | storage
+cat /proc/cmdline | storage
+echo "#####################################" | storage
 
-declare -a cpus
-cpus=(${cpulist})
+echo "**** uid: $UID ****" | storage
 
-for cmd in rtla; do
-    command -v $cmd >/dev/null 2>&1 || { echo >&2 "$cmd required but not installed.  Aborting"; exit 1; }
-done
+release=$(cat /etc/os-release | sed -n -r 's/VERSION_ID="(.).*/\1/p')
 
-cyccore=${cpus[1]}
-cindex=2
-ccount=1
-while (( $cindex < ${#cpus[@]} )); do
-    cyccore="${cyccore},${cpus[$cindex]}"
-    cindex=$(($cindex + 1))
-    ccount=$(($ccount + 1))
-done
+# Check the mode
+case "$COMMAND" in
+  "timerlat"|"hwnoise"|"osnoise")
+    echo "Operating in $COMMAND mode." | storage
+    ;;
+  *)
+    echo "Error: Invalid mode. Please set COMMAND to 'timerlat', 'hwnoise', or 'osnoise'." | storage
+    exit 1
+    ;;
+esac
+
+cpuset=`get_allowed_cpuset`
+echo "cpuset: $cpuset" | storage
 
-sibling=`cat /sys/devices/system/cpu/cpu${cpus[0]}/topology/thread_siblings_list | awk -F '[-,]' '{print $2}'`
-if [[ "${sibling}" =~ ^[0-9]+$ ]]; then
-    echo "removing cpu${sibling} from the cpu list because it is a sibling of cpu${cpus[0]} which will be the cpu-main-thread"
-    cyccore=${cyccore//,$sibling/}
+## ALLCPUSET handles a bug in rhel9 (and possible others) where cpuset is 0 when no cpuset is defined.
+## Deprecate this once the bug is fixed.
+if [[ "$cpuset" == "0" && "$ALLCPUSET" == "y" ]]; then
+    echo "present cpus: " | storage
+    work_cores=`get_host_cores`
+    echo "Info: No cpu-set used, using all CPUs" | storage
+    echo "WARNING: If trying to run on CPU0, make sure ALLCPUSET=n (Default behavior)" | storage
+else
+    if [[ "$cpuset" -eq 0 ]]; then
+        echo "Warning: cpuset is bound to CPU0. If you are trying to use all cpus add '-e ALLCPUSET=y' to your run command." | storage
+    fi
+    if [[ "$ALLCPUSET" == "y" ]]; then
+        echo "Warning: ALLCPUSET is set to y but cpuset was provided. ALLCPUSET will be ignored in favor of the provided cpuset." | storage
+    fi
+    work_cores=$cpuset
 fi
```

Comment:
I don't understand why this is necessary. I tested the existing script on OCP 4.13.9, which runs on RHEL 9.2 (5.14.0-284.25.1.rt14.310.el9_2.x86_64 kernel) and the cpuset is OK. What problem are you trying to fix?

Reply:
Maybe this was specific to my test machine. Im testing this as both a standalone container and an OCP container. Let me test this on a fresh instance. I wonder if there is a difference in how podman in configured on a standalone RHEL host vs coreos?

Reply:
The intended usecase for container-perf-tools is to be run from inside a pod on OCP (or possibly vanilla kubernetes). If you are running this in a standalone container, you would need to specify the cpus it should run on (e.g. in podman I think you can use the --cpuset-cpus option). We don't really test or support these tools from a standalone container, so some of them may not work at all.

Reply:
This bug has since been fixed in rhel9. So Ill remove this.